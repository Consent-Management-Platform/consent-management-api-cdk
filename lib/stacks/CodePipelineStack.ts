import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Effect, OpenIdConnectPrincipal, OpenIdConnectProvider, PolicyDocument, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

import { GITHUB_ORGANIZATION } from '../constants/github';
import { StageConfig } from '../interfaces/stage-config';

export interface CodePipelineStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines infrastructure supporting automated deployment pipelines.
 */
export class CodePipelineStack extends Stack {
  constructor(scope: Construct, id: string, readonly props: CodePipelineStackProps) {
    super(scope, id, props);

    // Create GitHub OpenID Connect provider
    const oidcProvider = new OpenIdConnectProvider(this, 'GitHubOIDCProvider', {
      url: 'https://token.actions.githubusercontent.com',
      clientIds: ['sts.amazonaws.com'],
    });

    // Create GitHub principal which will be used to assume deployment roles.
    const gitHubPrincipal = new OpenIdConnectPrincipal(oidcProvider).withConditions(
      {
        StringLike: {
          'token.actions.githubusercontent.com:sub':
            `repo:${GITHUB_ORGANIZATION}/*:*`,
        },
      }
    );

    /**
      * Create a deployment role that has short lived credentials. The only
      * principal that can assume this role is the GitHub Open ID provider.
      *
      * This role is granted authority to assume AWS CDK roles.
      */
    new Role(this, 'GitHubActionsRole', {
      assumedBy: gitHubPrincipal,
      description: 'Role assumed by GitHub Actions to deploy code/infra using aws cdk',
      roleName: 'github-ci-role',
      maxSessionDuration: Duration.hours(1),
      inlinePolicies: {
        CdkDeploymentPolicy: new PolicyDocument({
          assignSids: true,
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['sts:AssumeRole'],
              resources: [`arn:aws:iam::${this.account}:role/cdk-*`],
            }),
          ],
        }),
      },
    });
  }
}
