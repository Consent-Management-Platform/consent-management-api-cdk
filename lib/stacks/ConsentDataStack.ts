import { Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, BillingMode, ProjectionType, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

import { StageConfig } from '../interfaces/stage-config';

export interface ConsentDataStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines the consent data storage layer.
 *
 * This stack is separated from higher-level components such as API compute services
 * to minimize the risk of other layer changes impacting the consent database,
 * which should be extremely stable and never deleted after prod launch.
 */
export class ConsentDataStack extends Stack {
  public readonly consentTable: Table;

  constructor(scope: Construct, id: string, readonly props: ConsentDataStackProps) {
    super(scope, id, props);

    this.consentTable = this.createConsentTable();
    this.createConsentsByServiceUserGsi(this.consentTable);
  }

  private createConsentTable() {
    return new Table(this, 'ServiceUserConsentDynamoDBTable', {
      tableName: 'ServiceUserConsent',
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: {
        name: 'Id',
        type: AttributeType.STRING
      },
      pointInTimeRecovery: true,
      deletionProtection: true,
      billingMode: BillingMode.PAY_PER_REQUEST
    });
  }

  private createConsentsByServiceUserGsi(consentTable: Table) {
    consentTable.addGlobalSecondaryIndex({
      indexName: 'ConsentsByServiceUser',
      partitionKey: {
        name: 'UserId',
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'ServiceId',
        type: AttributeType.STRING
      },
      projectionType: ProjectionType.ALL
    });
  }
}
