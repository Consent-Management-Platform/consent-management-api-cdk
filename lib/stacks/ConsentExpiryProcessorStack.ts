import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Table } from 'aws-cdk-lib/aws-dynamodb';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Code, Function, Runtime, SnapStartConf } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

import { StageConfig } from '../interfaces/stage-config';
import { CustomLambdaFunction } from '../constructs/CustomLambdaFunction';
import { Rule, Schedule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction } from 'aws-cdk-lib/aws-events-targets';

export interface ConsentExpiryProcessorStackProps extends StackProps {
  codePackageFilePath: string;
  consentTable: Table;
  stageConfig: StageConfig;
}

/**
 * Defines infrastructure for automatically updating consents to EXPIRED when they have passed their expiry time.
 */
export class ConsentExpiryProcessorStack extends Stack {
  public readonly expiryProcessorLambda: Function;

  constructor(scope: Construct, id: string, readonly props: ConsentExpiryProcessorStackProps) {
    super(scope, id, props);

    this.expiryProcessorLambda = this.createExpiryProcessorLambdaFunction();
    this.createLambdaSchedule(this.expiryProcessorLambda);
  }

  private createExpiryProcessorLambdaFunction(): Function {
    const lambdaFunction: Function = new CustomLambdaFunction(this, 'ConsentExpiryProcessorLambda', {
      code: Code.fromAsset(this.props.codePackageFilePath),
      description: 'Consent Expiry Processor Lambda',
      handler: 'com.consentframework.consentexpiryprocessor.ConsentExpiryProcessor::handleRequest',
      memorySize: 1536,
      runtime: Runtime.JAVA_21,
      snapStart: SnapStartConf.ON_PUBLISHED_VERSIONS,
      timeout: Duration.minutes(15)
    });

    // Grant permissions to query DynamoDB consent data
    lambdaFunction.addToRolePolicy(new PolicyStatement({
      sid: 'ConsentDynamoDBQueryPermissions',
      actions: [
        'dynamodb:ConditionCheckItem',
        'dynamodb:Query',
        'dynamodb:Scan',
        'dynamodb:UpdateItem'
      ],
      resources: [
        this.props.consentTable.tableArn,
        `${this.props.consentTable.tableArn}/index/*`
      ]
    }));

    return lambdaFunction;
  }

  // Schedule the Lambda function to run hourly
  private createLambdaSchedule(apiLambda: Function) {
    // Choose a random minute and distribute with any other scheduled jobs to avoid load spikes
    const scheduleRule = new Rule(this, 'ConsentExpiryProcessorSchedule', {
      schedule: Schedule.cron({
        minute: '51'
      })
    });

    scheduleRule.addTarget(new LambdaFunction(apiLambda));
  }
}
