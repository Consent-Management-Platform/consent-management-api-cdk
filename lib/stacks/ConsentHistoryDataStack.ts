import { Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, BillingMode, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

import { StageConfig } from '../interfaces/stage-config';

export interface ConsentHistoryDataStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines the consent history data storage layer.
 */
export class ConsentHistoryDataStack extends Stack {
  public readonly consentHistoryTable: Table;

  constructor(scope: Construct, id: string, readonly props: ConsentHistoryDataStackProps) {
    super(scope, id, props);

    this.consentHistoryTable = this.createConsentHistoryTable();
  }

  private createConsentHistoryTable() {
    return new Table(this, 'ServiceUserConsentHistoryDynamoDBTable', {
      tableName: 'ConsentHistory',
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'eventId',
        type: AttributeType.STRING
      },
      pointInTimeRecovery: true,
      deletionProtection: true,
      billingMode: BillingMode.PAY_PER_REQUEST
    });
  }
}
