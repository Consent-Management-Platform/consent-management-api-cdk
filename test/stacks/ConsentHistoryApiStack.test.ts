import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { join } from 'path';

import { MOCK_ENV } from '../../fixtures/mock-env';
import { MOCK_STAGE_CONFIG } from '../../fixtures/mock-stage-config';
import { CodePipelineStack } from '../../lib/stacks/CodePipelineStack';
import { ConsentHistoryDataStack } from '../../lib/stacks/ConsentHistoryDataStack';
import { ConsentHistoryApiStack } from '../../lib/stacks/ConsentHistoryApiStack';

describe('ConsentHistoryApiStack', () => {
  // Normalize template IDs and attributes that change based on upstream package updates
  function cleanApiResources(templateJson: any) {
    const API_DEPLOYMENT_KEY_PREFIX = 'ConsentHistoryApiDeployment';
    const API_LAMBDA_INVOKE_KEY_PREFIX = 'ConsentHistoryApiLambdaInvoke';
    const API_GATEWAY_DEPLOYMENT_KEY_PREFIX = 'ConsentHistoryAPIGatewayDeployment';
    const API_GATEWAY_DEPLOYMENT_STAGE_KEY_PREFIX = 'ConsentHistoryAPIGatewayDeploymentStageprod';
    const OUTPUTS_KEY = 'Outputs';
    const PROPERTIES_KEY = 'Properties';
    const RESOURCES_KEY = 'Resources';

    // Remove API service code S3 path references which change on every API service package update
    delete templateJson[RESOURCES_KEY]['ConsentHistoryApiLambda03FB76FC'][PROPERTIES_KEY]['Code'];

    // Normalize resource keys which change on upstream package updates
    for (const resourceKey in templateJson[RESOURCES_KEY]) {
      if (resourceKey.startsWith(API_LAMBDA_INVOKE_KEY_PREFIX)) {
        templateJson[RESOURCES_KEY][API_LAMBDA_INVOKE_KEY_PREFIX] = templateJson[RESOURCES_KEY][resourceKey];
        delete templateJson[RESOURCES_KEY][resourceKey];
      } else if (resourceKey.startsWith(API_GATEWAY_DEPLOYMENT_STAGE_KEY_PREFIX)) {
        templateJson[RESOURCES_KEY][API_GATEWAY_DEPLOYMENT_STAGE_KEY_PREFIX] = templateJson[RESOURCES_KEY][resourceKey];
        delete templateJson[RESOURCES_KEY][API_GATEWAY_DEPLOYMENT_STAGE_KEY_PREFIX][PROPERTIES_KEY]['DeploymentId'];
        delete templateJson[RESOURCES_KEY][resourceKey];
      } else if (resourceKey.startsWith(API_GATEWAY_DEPLOYMENT_KEY_PREFIX)) {
        templateJson[RESOURCES_KEY][API_GATEWAY_DEPLOYMENT_KEY_PREFIX] = templateJson[RESOURCES_KEY][resourceKey];
        delete templateJson[RESOURCES_KEY][resourceKey];
      } else if (resourceKey.startsWith(API_DEPLOYMENT_KEY_PREFIX)) {
        delete templateJson[RESOURCES_KEY][resourceKey];
      }
    }

    // Strip outputs that change dependent on upstream package updates
    for (const outputKey in templateJson[OUTPUTS_KEY]) {
      if (outputKey.startsWith('ConsentHistoryAPIGatewayEndpoint')) {
        delete templateJson[OUTPUTS_KEY][outputKey];
      }
    }
  }

  it('creates the expected CloudFormation template from CDK', () => {
    const app = new App();

    const codePipelineStack = new CodePipelineStack(app, 'CodePipelineStack', {
      env: MOCK_ENV,
      stageConfig: MOCK_STAGE_CONFIG
    });
    const dataStack = new ConsentHistoryDataStack(app, 'ConsentHistoryDataStack', {
      env: MOCK_ENV,
      stageConfig: MOCK_STAGE_CONFIG
    });
    const apiStack = new ConsentHistoryApiStack(app, 'ConsentHistoryApiStack', {
      env: MOCK_ENV,
      stageConfig: MOCK_STAGE_CONFIG,
      apiCodePackageFilePath: join(__dirname, '../../../consent-history-api'),
      codeDeployRole: codePipelineStack.codeDeployRole,
      consentHistoryTable: dataStack.consentHistoryTable
    });

    const templateJson = Template.fromStack(apiStack).toJSON();
    cleanApiResources(templateJson);

    expect(templateJson).toMatchSnapshot();
  });
});
