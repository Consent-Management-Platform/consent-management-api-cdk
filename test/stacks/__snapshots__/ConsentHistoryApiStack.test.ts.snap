// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsentHistoryApiStack creates the expected CloudFormation template from CDK 1`] = `
{
  "Outputs": {
    "ConsentHistoryApiGatewayIdExport": {
      "Export": {
        "Name": "ConsentHistoryApiGatewayId",
      },
      "Value": {
        "Ref": "ConsentHistoryAPIGateway7E4863E3",
      },
    },
    "ExportsOutputRefConsentHistoryAPIGateway7E4863E3D4251513": {
      "Export": {
        "Name": "ConsentHistoryApiStack:ExportsOutputRefConsentHistoryAPIGateway7E4863E3D4251513",
      },
      "Value": {
        "Ref": "ConsentHistoryAPIGateway7E4863E3",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ConsentHistoryAPIGateway7E4863E3": {
      "Properties": {
        "Body": {
          "components": {
            "schemas": {
              "Consent": {
                "description": "Record of user consent.",
                "properties": {
                  "consentData": {
                    "$ref": "#/components/schemas/ConsentData",
                  },
                  "consentId": {
                    "description": "Unique identifier for a Consent.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "consentType": {
                    "description": "Consent type.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "consentVersion": {
                    "type": "integer",
                  },
                  "expiryTime": {
                    "description": "Time when the Consent automatically expires, using the ISO 8601 format.",
                    "format": "date-time",
                    "type": "string",
                  },
                  "serviceId": {
                    "description": "Unique identifier for a service.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "status": {
                    "$ref": "#/components/schemas/ConsentStatus",
                  },
                  "userId": {
                    "description": "Unique identifier for a user.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                },
                "required": [
                  "consentId",
                  "consentVersion",
                  "serviceId",
                  "status",
                  "userId",
                ],
                "type": "object",
              },
              "ConsentChangeEvent": {
                "description": "A user consent change event.",
                "properties": {
                  "consentId": {
                    "description": "Unique identifier for a Consent.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "eventId": {
                    "description": "Unique identifier for a consent change event.",
                    "type": "string",
                  },
                  "eventTime": {
                    "description": "Consent event timestamp, using the ISO 8601 format.",
                    "format": "date-time",
                    "type": "string",
                  },
                  "eventType": {
                    "$ref": "#/components/schemas/ConsentEventType",
                  },
                  "newImage": {
                    "$ref": "#/components/schemas/Consent",
                  },
                  "oldImage": {
                    "$ref": "#/components/schemas/Consent",
                  },
                },
                "required": [
                  "consentId",
                  "eventId",
                  "eventTime",
                  "eventType",
                ],
                "type": "object",
              },
              "ConsentData": {
                "additionalProperties": {
                  "maxLength": 256,
                  "minLength": 1,
                  "type": "string",
                },
                "description": "Consent data represented by string key-value pairs. Keys must be between 1-64 characters long, values must be between 1-256 characters long, and up to 32 key-value pairs may be provided.",
                "maxProperties": 32,
                "type": "object",
              },
              "ConsentEventType": {
                "description": "Type of consent change event.",
                "enum": [
                  "INSERT",
                  "MODIFY",
                  "REMOVE",
                ],
                "type": "string",
              },
              "ConsentHistory": {
                "description": "A specific consent's ID and associated consent change events.",
                "properties": {
                  "consentId": {
                    "description": "Unique identifier for a Consent.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "history": {
                    "description": "List of user consent change events.",
                    "items": {
                      "$ref": "#/components/schemas/ConsentChangeEvent",
                    },
                    "type": "array",
                  },
                },
                "required": [
                  "consentId",
                ],
                "type": "object",
              },
              "ConsentStatus": {
                "description": "Current status of a Consent.",
                "enum": [
                  "ACTIVE",
                  "EXPIRED",
                  "REVOKED",
                ],
                "type": "string",
              },
              "GetHistoryForServiceUserConsentResponseContent": {
                "properties": {
                  "data": {
                    "description": "List of user consent change events.",
                    "items": {
                      "$ref": "#/components/schemas/ConsentChangeEvent",
                    },
                    "type": "array",
                  },
                },
                "type": "object",
              },
              "GetHistoryForServiceUserResponseContent": {
                "properties": {
                  "data": {
                    "description": "List of consent histories.",
                    "items": {
                      "$ref": "#/components/schemas/ConsentHistory",
                    },
                    "type": "array",
                  },
                },
                "type": "object",
              },
              "InternalServiceExceptionResponseContent": {
                "description": "This exception is thrown on an unhandled service error.",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "NotFoundExceptionResponseContent": {
                "description": "This exception is thrown when the requested entity is not found.",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "type": "object",
              },
            },
          },
          "info": {
            "title": "Consent History API",
            "version": "2025-04-06",
          },
          "openapi": "3.0.2",
          "paths": {
            "/v1/consent-history/services/{serviceId}/users/{userId}/consents": {
              "get": {
                "description": "Retrieve history for a given service user.",
                "operationId": "GetHistoryForServiceUser",
                "parameters": [
                  {
                    "description": "Unique identifier for a service.",
                    "in": "path",
                    "name": "serviceId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a service.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a user.",
                    "in": "path",
                    "name": "userId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a user.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                ],
                "responses": {
                  "200": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/GetHistoryForServiceUserResponseContent",
                        },
                      },
                    },
                    "description": "GetHistoryForServiceUser 200 response",
                  },
                  "404": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/NotFoundExceptionResponseContent",
                        },
                      },
                    },
                    "description": "NotFoundException 404 response",
                  },
                  "500": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/InternalServiceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "InternalServiceException 500 response",
                  },
                },
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:apigateway:us-east-2:lambda:path/2015-03-31/functions/",
                        {
                          "Fn::GetAtt": [
                            "ConsentHistoryApiLambda03FB76FC",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
            },
            "/v1/consent-history/services/{serviceId}/users/{userId}/consents/{consentId}": {
              "get": {
                "description": "Retrieve history for a given service user consent.",
                "operationId": "GetHistoryForServiceUserConsent",
                "parameters": [
                  {
                    "description": "Unique identifier for a Consent.",
                    "in": "path",
                    "name": "consentId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a Consent.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a service.",
                    "in": "path",
                    "name": "serviceId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a service.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a user.",
                    "in": "path",
                    "name": "userId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a user.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                ],
                "responses": {
                  "200": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/GetHistoryForServiceUserConsentResponseContent",
                        },
                      },
                    },
                    "description": "GetHistoryForServiceUserConsent 200 response",
                  },
                  "404": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/NotFoundExceptionResponseContent",
                        },
                      },
                    },
                    "description": "NotFoundException 404 response",
                  },
                  "500": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/InternalServiceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "InternalServiceException 500 response",
                  },
                },
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:apigateway:us-east-2:lambda:path/2015-03-31/functions/",
                        {
                          "Fn::GetAtt": [
                            "ConsentHistoryApiLambda03FB76FC",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
            },
          },
          "x-amazon-apigateway-request-validator": "full",
          "x-amazon-apigateway-request-validators": {
            "full": {
              "validateRequestBody": true,
              "validateRequestParameters": true,
            },
          },
        },
        "EndpointConfiguration": {
          "Types": [
            "EDGE",
          ],
        },
        "Name": "ConsentHistoryApi",
        "Policy": {
          "Statement": [
            {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
              "Principal": {
                "AWS": [
                  {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::123456789012:root",
                      ],
                    ],
                  },
                  {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:iam::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":role/github-ci-role",
                      ],
                    ],
                  },
                ],
              },
              "Resource": "execute-api/*",
              "Sid": "ConsentHistoryApiGatewayFullInvokeAccess",
            },
            {
              "Action": [
                "apigateway:GET",
                "apigateway:POST",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":role/github-ci-role",
                    ],
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":apigateway:",
                      {
                        "Ref": "AWS::Region",
                      },
                      "::/restapis/*/resources",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":apigateway:",
                      {
                        "Ref": "AWS::Region",
                      },
                      "::/restapis/*/resources/*/*/*",
                    ],
                  ],
                },
              ],
              "Sid": "ConsentHistoryApiGatewayApiAccess",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ConsentHistoryAPIGatewayAccount26AE8621": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "ConsentHistoryAPIGateway7E4863E3",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "ConsentHistoryAPIGatewayCloudWatchRoleE83540C9",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ConsentHistoryAPIGatewayCloudWatchRoleE83540C9": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ConsentHistoryAPIGatewayDeployment": {
      "Properties": {
        "Description": "Consent History API, see documentation at https://consent-management-platform.github.io/consent-history-api-models/v1/docs.html",
        "RestApiId": {
          "Ref": "ConsentHistoryAPIGateway7E4863E3",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ConsentHistoryAPIGatewayDeploymentStageprod": {
      "DependsOn": [
        "ConsentHistoryAPIGatewayAccount26AE8621",
      ],
      "Properties": {
        "MethodSettings": [
          {
            "DataTraceEnabled": false,
            "HttpMethod": "*",
            "LoggingLevel": "INFO",
            "MetricsEnabled": true,
            "ResourcePath": "/*",
            "ThrottlingBurstLimit": 5,
            "ThrottlingRateLimit": 3,
          },
        ],
        "RestApiId": {
          "Ref": "ConsentHistoryAPIGateway7E4863E3",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ConsentHistoryApiLambda03FB76FC": {
      "DependsOn": [
        "ConsentHistoryApiLambdaServiceRoleDefaultPolicy5F5E8C68",
        "ConsentHistoryApiLambdaServiceRoleA0D24C3A",
      ],
      "Properties": {
        "Description": "Consent History API Lambda",
        "Handler": "com.consentframework.consenthistory.api.ConsentHistoryApiService::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "ConsentHistoryApiLambdaLogGroup4D13728C",
          },
        },
        "MemorySize": 1536,
        "Role": {
          "Fn::GetAtt": [
            "ConsentHistoryApiLambdaServiceRoleA0D24C3A",
            "Arn",
          ],
        },
        "Runtime": "java21",
        "SnapStart": {
          "ApplyOn": "PublishedVersions",
        },
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConsentHistoryApiLambdaInvoke": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ConsentHistoryApiLambda03FB76FC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-2:123456789012:",
              {
                "Ref": "ConsentHistoryAPIGateway7E4863E3",
              },
              "/*/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ConsentHistoryApiLambdaLogGroup4D13728C": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "ConsentHistoryApiLambda-ApplicationLogs",
        "RetentionInDays": 545,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "ConsentHistoryApiLambdaServiceRoleA0D24C3A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ConsentHistoryApiLambdaServiceRoleDefaultPolicy5F5E8C68": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "CloudWatchMetricsPermissions",
            },
            {
              "Action": [
                "dynamodb:ConditionCheckItem",
                "dynamodb:GetItem",
                "dynamodb:Query",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "ConsentHistoryDataStack:ExportsOutputFnGetAttServiceUserConsentHistoryDynamoDBTable552A6381Arn5A6BE492",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "ConsentHistoryDataStack:ExportsOutputFnGetAttServiceUserConsentHistoryDynamoDBTable552A6381Arn5A6BE492",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
              "Sid": "ConsentHistoryDdbQueryPermissions",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ConsentHistoryApiLambdaServiceRoleDefaultPolicy5F5E8C68",
        "Roles": [
          {
            "Ref": "ConsentHistoryApiLambdaServiceRoleA0D24C3A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
